### PR TITLE
[ci full] Fix #3766: high level content signature chain of trust verification

### LINKED ISSUE #3766: Expose NSS VerifyCodeSigningCertificateChain in `rc_crypto` (requires NSS 3.66)
Periodic update.  As part of the week, need to review the existing documentation.

The rc_crypto crate is probably our scariest crate, both because it's crypto-related and because it's got some weird licensing stuff going on. It's been a while since any of us had to deal with said scariness, so it's a good time to revisit that crate with fresh eyes and see if there are any docs that need to be updated (or brought into existence).

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-2586)
┆Epic: a-s supporting tools (backlog)

### PR DESCRIPTION
**Pull request by leplatrem**
Fix #3766

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

**Comment by leplatrem** [comment]
I'm stuck on this link error:
```
/usr/bin/ld: /home/mathieu/Code/Mozilla/application-services/target/debug/deps/systest-5615b05b28b8b37e.4ehwytlossh7jkgj.rcgu.o: in function `systest::fn_VerifyCodeSigningCertificateChain':
          /home/mathieu/Code/Mozilla/application-services/target/debug/build/systest-6de8ac3438657825/out/all.rs:(.text._ZN7systest36fn_VerifyCodeSigningCertificateChain17h06fc7a56f4e12328E+0xe): undefined reference to `VerifyCodeSigningCertificateChain'

```

Whereas I can see that `"-L" "/home/mathieu/Code/Mozilla/application-services/libs/desktop/linux-x86-64/nss/lib"` is passed to the compilation command, and that `nm libs/desktop/linux-x86-64/nss/lib/libmozpkix.a` shows `VerifyCodeSigningCertificateChain`.

@rfk do you have any idea?

**Comment by leplatrem** [comment]
>  @rfk do you have any idea?

Dana unblocked me :) I also added a flag to `stdc++`, I don't know what it implies elsewhere in the bootstrapping area.

Basically, my next steps are:
- expose `CERT_GetDefaultCertDB()`
- expose `CERT_NewTempCertificate()` and `CERT_ExtractPublicKey()`
- use them to parse the leaf certificate, and extract its public key, to verify the content signature, using the code that I had in remote-settings-client already

File: components/support/rc_crypto/nss/nss_build_common/src/lib.rs
Comment (rfk) [inline components/support/rc_crypto/nss/nss_build_common/src/lib.rs:82]: Nice find. I would like us to be using a more structured approach to the build here such as the [`cc` crate](https://github.com/alexcrichton/cc-rs) but that's well beyond the scope of this PR.

One thing to watch out for here, is apparently the name of this may change on different platforms, as described [here](https://github.com/alexcrichton/cc-rs#c-support). We'll have to do some manual testing on Android to see if any tweaks are needed.

Comment (rfk) [reply components/support/rc_crypto/nss/nss_build_common/src/lib.rs:82]: On my Mac, I had to change this to `println!("cargo:rustc-link-lib=c++");` to get it to compile successfully. (The tests fail in something about splitting the pem, but that seems unlikely to be related to c++ linking issues).

File: components/support/rc_crypto/nss/src/pkixc.rs
Comment (rfk) [inline components/support/rc_crypto/nss/src/pkixc.rs:37]: Careful, you haven't checked the length of `root_sha256_hash` anywhere here. For safety, we should either check it here in the function body, or insist on it in the type of the parameter, which IIRC would be something like `root_sha256_hash: &[u8; 32]`. Which approach is better probably depends on the details of how it's being called.

(I see that you check for this in the higher-level code, but it's good practice to put these checks next to the `unsafe` code that depends on them, otherwise we're pretending that `verify_code_signing_certificate_chain` is safe when it isn't really).

**Comment by leplatrem** [comment]
I'm not sure how to fix the failing tests, but at last the code is ready to be reviewed :)

Thanks :pray:

**Comment by rfk** [comment]
@leplatrem thanks! I'm keen to dive in here, but also I have a week of PTO after the wellness week so I may not get a chance to look before the week of 12th July. Do you need this looked at before then?

**Comment by leplatrem** [comment]
@rfk Thanks! Don't worry, more or less the same here :)

**Comment by rfk** [comment]
Sorry for the ongoing delay here, I've been caught up in a few other things after turning from break. Please hassle me if I haven't commented substantially here by the end of the week.

File: components/support/rc_crypto/nss/src/pkixc.rs
Comment (mozkeeler) [inline components/support/rc_crypto/nss/src/pkixc.rs:16]: https://searchfox.org/mozilla-central/rev/04b41051dd2b08df2a06f86f8bf9230c0848fc51/security/nss/lib/mozpkix/include/pkix/pkixnss.h#81

File: components/support/rc_crypto/nss/src/pkixc.rs
Comment (mozkeeler) [inline components/support/rc_crypto/nss/src/pkixc.rs:15]: https://searchfox.org/mozilla-central/rev/04b41051dd2b08df2a06f86f8bf9230c0848fc51/security/nss/lib/ssl/sslerr.h#42

**Comment by rfk** [comment]
@leplatrem I think rebasing should fix the failing "rust dependencies" test. I'll see if I can reproduce the carthage build failures locally on my mac.

File: components/support/rc_crypto/nss/nss_sys/src/bindings/pkixc.rs
Comment (rfk) [inline components/support/rc_crypto/nss/nss_sys/src/bindings/pkixc.rs:27]: nit: please move this to above the function that it documents, for consistency with the other file.

File: components/support/rc_crypto/nss/src/cert.rs
Comment (rfk) [inline components/support/rc_crypto/nss/src/cert.rs:7]: nit: remove this if unneeded

File: components/support/rc_crypto/nss/src/cert.rs
Comment (rfk) [inline components/support/rc_crypto/nss/src/cert.rs:36]: Yes please :-)

File: components/support/rc_crypto/nss/src/pkixc.rs
Comment (rfk) [inline components/support/rc_crypto/nss/src/pkixc.rs:30]: Is it worth erroring out here if the length is too large to fit in a `u16`? IIUC Rust will just truncate the bits when casting to a smaller integer size, which will always produce a smaller length, which will not risk memory unsafety...so I guess it's fine? May be worth a comment to that effect at least.

File: components/support/rc_crypto/src/contentsignature.rs
Comment (rfk) [inline components/support/rc_crypto/src/contentsignature.rs:4]: Please include some brief high-level module docs here, even if just some links out to further reading. I fear future maintainers of this crate may not have all the context to remember what this is for.

File: components/support/rc_crypto/src/contentsignature.rs
Comment (rfk) [inline components/support/rc_crypto/src/contentsignature.rs:11]: An example of the sort of string we're trying to pass here would be useful to help understanding, I am interpreting it as something like "ab:12:ef:...".

File: components/support/rc_crypto/src/contentsignature.rs
Comment (rfk) [inline components/support/rc_crypto/src/contentsignature.rs:20]: If you wanted to mess around with Rust iterators, this would be a good opportunity to use `collect`. It's possible to [`collect` into a `Result<Vec<T>>`](https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect) and get the early-termination-on-error thing for free, something along the lines of:

```
fn decode_root_hash(input: &str) -> Result<Vec<u8>> {
    input.split(':').map(|byte_hex| {
        hex::decode(&byte_hex).map_err(|_| {
            ErrorKind::RootHashFormatError(input.to_string()).into()
        })
    }).collect()
}
```

(The more explicit version is also fine if you *don't* want to mess around with Rust iterators).

File: components/support/rc_crypto/src/contentsignature.rs
Comment (rfk) [inline components/support/rc_crypto/src/contentsignature.rs:25]: Please add docs here to help understand the `Vec<Vec<u8>>` situation

File: components/support/rc_crypto/src/contentsignature.rs
Comment (rfk) [inline components/support/rc_crypto/src/contentsignature.rs:73]: You might be able to use something like `let mut certificate_slices = certificates.iter().collect();` here for succinctness, letting Rust infer the type based on its future use as a parameter to `verify_code_signing_certificate_chain`.

File: components/support/rc_crypto/src/contentsignature.rs
Comment (rfk) [inline components/support/rc_crypto/src/contentsignature.rs:105]: Out of curiosity, what would happen if we *didn't* do these checks?

File: components/support/rc_crypto/src/contentsignature.rs
Comment (rfk) [inline components/support/rc_crypto/src/contentsignature.rs:283]: It would be great to give each of these tests a comment to indicate the specific reason why the verification fails, e.g. "Fails with invalid hostname", "Fails with expired signature", etc. Also it would be great to have tests for a few different ways it can fail - not to check that NSS is doing what we expect, but to ensure that we're not e.g. accidentally always passing an empty string or a zero timestamp or something.

**Review by rfk — COMMENTED** [review]
This looks pretty good to me, although fiddling around in the low-level NSS bits always makes me nervous. Next steps here:

* I have the tests passing locally on my mac; I need to try to reproduce the CI *failures* and help you debug from there.
* I left a couple of comments of places where we could use some of Rust's more advanced features to make the code more succinct, but up to you if you want to take them.
* I think we need to add a few more tests and some docs before final merge.

I'm pretty excited to see this coming together!

**Comment by rfk** [comment]
> I have the tests passing locally on my mac; I need to try to reproduce the CI failures and help you debug from there.

A quick google suggests this is *probably* to do with linking against the C++ stdlib from within an XCode build, e.g. https://github.com/auth0/react-native-lock/issues/88. So hopefully it won't be too thorny to get to the bottom of...

File: components/support/rc_crypto/src/contentsignature.rs
Comment (leplatrem) [reply components/support/rc_crypto/src/contentsignature.rs:73]: I think I've tried and got out of compilation errors with the long version...

```rust
79 |     let mut certificates_slices = certificates.iter().collect();
   |                                                       ^^^^^^^ value of type `Vec<&[u8]>` cannot be built from `std::iter::Iterator<Item=&Vec<u8>>`
```

File: components/support/rc_crypto/src/contentsignature.rs
Comment (leplatrem) [reply components/support/rc_crypto/src/contentsignature.rs:20]: I could not figure out how to flatten items and collect results, so I ended up writing an explicit loop :man_shrugging:

**Comment by leplatrem** [comment]
Thanks @rfk @mozkeeler for your review and help! Tests now pass :tada: 

I added a few commits, I'd be happy to rebase/reorganize them once reviewed. 

Please let me know if you expect more test cases for the verification. If we want to recreate invalid certificates chains for different fined-grained reasons, I may have to look for tools or solutions to alter existing valid certs...
I also added 2 new kinds of errors (SSL and PKIX) that would help find out the original NSS error from the integer code more easily (IMO).

File: components/support/rc_crypto/src/contentsignature.rs
Comment (leplatrem) [reply components/support/rc_crypto/src/contentsignature.rs:123]: I added a comment below, but basically the signature verification would fail for a number of different reasons. Doing these checks here help us being more explicit about the reason of failure.

**Review by mozkeeler — CHANGES_REQUESTED** [review]
This looks really good. I mostly looked at the new NSS parts and noted a few changes that should be made.

File: components/support/rc_crypto/nss/src/cert.rs
Comment (rfk) [reply components/support/rc_crypto/nss/src/cert.rs:39]: Good call, I believe we have a `ScopedPtr` wrapper struct that is used to help with this in other functions. See `src/pk11/types.rs` for where it's used to define RAII wrappers on some other pointer types.

In fact, at least `SECKEYPublicKey` seems to already have a `ScopedPtr` wrapper struct. You may be able to achieve both the null-check and the automatic cleanup by doing something like:

```
use crate::pk11::types::PublicKey;
[...]
let pub_key = unsafe { PublicKey::from_ptr(CERT_ExtractPublicKey(cert)) }
```

The `CERTCertificate` types will need to be added to the list in `types.rs` to achieve a similar affect.

File: components/support/rc_crypto/src/contentsignature.rs
Comment (rfk) [reply components/support/rc_crypto/src/contentsignature.rs:29]: ah well, it was worth a try :-)

**Review by rfk — CHANGES_REQUESTED** [review]
The higher-level `rc_crypto` bits here look good to me, thanks!

@mozkeeler's suggestion of an RAII guard on the low-level NSS types is spot on and hopefully fairly straightforward using the `scoped_ptr!` macro that I mention below, and I think it's worth me taking another look at this once that's in place, just for completeness.

File: components/support/rc_crypto/nss/src/pkixc.rs
Comment (leplatrem) [reply components/support/rc_crypto/nss/src/pkixc.rs:69]: I'm not really aware of the implications of this here, since Rust input parameters are not mutable anyway. But since it's a mistake on the exposed API I'll add a comment

**Review by rfk — APPROVED** [review]
Nice, the scoped-ptr stuff looks 👍🏻

**Comment by rfk** [comment]
I think this is good to go when you're happy with it @leplatrem. Please consider a bit of a rebase and history cleanup, then add the `checkin-needed` label when it's ready to land. (I'm not sure if the bot will successfully do its thing in that case, but it will at the very least signal to us that it's ready for merge).

**Pull request merged**
